### ISSUE TITLE
notebook errors with unable to create directory

### ISSUE THREAD

**Issue by satra** [jupyter/notebook#1411]
error details, below on a clean environment:

```
$ jupyter-notebook --debug
[D 09:59:58.787 NotebookApp] Searching [u'/om/user/satra/projects/SAD/scripts/fmri', '/home/satra/.jupyter', '/om/user/satra/envs/testjupyter/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files  
[D 09:59:58.788 NotebookApp] Looking for jupyter_config in /etc/jupyter
[D 09:59:58.789 NotebookApp] Looking for jupyter_config in /usr/local/etc/jupyter
[D 09:59:58.789 NotebookApp] Looking for jupyter_config in /om/user/satra/envs/testjupyter/etc/jupyter  
[D 09:59:58.789 NotebookApp] Looking for jupyter_config in /home/satra/.jupyter
[D 09:59:58.790 NotebookApp] Looking for jupyter_config in /om/user/satra/projects/SAD/scripts/fmri
[D 09:59:58.791 NotebookApp] Looking for jupyter_notebook_config in /etc/jupyter
[D 09:59:58.791 NotebookApp] Looking for jupyter_notebook_config in /usr/local/etc/jupyter
[D 09:59:58.792 NotebookApp] Looking for jupyter_notebook_config in /om/user/satra/envs/testjupyter/etc/jupyter
[D 09:59:58.792 NotebookApp] Looking for jupyter_notebook_config in /home/satra/.jupyter
[D 09:59:58.792 NotebookApp] Looking for jupyter_notebook_config in /om/user/satra/projects/SAD/scripts/fmri
Traceback (most recent call last):
  File "/om/user/satra/envs/testjupyter/bin/jupyter-notebook", line 6, in <module>
    sys.exit(notebook.notebookapp.main())
  File "/om/user/satra/envs/testjupyter/lib/python2.7/site-packages/jupyter_core/application.py", line 267, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/om/user/satra/envs/testjupyter/lib/python2.7/site-packages/traitlets/config/application.py", line 595, in launch_instance
    app.initialize(argv)
  File "<decorator-gen-7>", line 2, in initialize
  File "/om/user/satra/envs/testjupyter/lib/python2.7/site-packages/traitlets/config/application.py", line 74, in catch_config_error
    return method(app, *args, **kwargs)
  File "/om/user/satra/envs/testjupyter/lib/python2.7/site-packages/notebook/notebookapp.py", line 1058, in initialize
    self.init_configurables()
  File "/om/user/satra/envs/testjupyter/lib/python2.7/site-packages/notebook/notebookapp.py", line 838, in init_configurables
    connection_dir=self.runtime_dir,
  File "/om/user/satra/envs/testjupyter/lib/python2.7/site-packages/traitlets/traitlets.py", line 529, in __get__
    return self.get(obj, cls)
  File "/om/user/satra/envs/testjupyter/lib/python2.7/site-packages/traitlets/traitlets.py", line 508, in get
    value = self._validate(obj, dynamic_default())
  File "/om/user/satra/envs/testjupyter/lib/python2.7/site-packages/jupyter_core/application.py", line 99, in _runtime_dir_default
    ensure_dir_exists(rd, mode=0o700)
  File "/om/user/satra/envs/testjupyter/lib/python2.7/site-packages/ipython_genutils/path.py", line 167, in ensure_dir_exists
    os.makedirs(path, mode=mode)
  File "/om/user/satra/envs/testjupyter/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/om/user/satra/envs/testjupyter/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 13] Permission denied: '/run/user/48529'
```

environment details, on centos 7.2, using a slurm terminal:
```
name: testjupyter
dependencies:
- backports=1.0=py27_0
- backports_abc=0.4=py27_0
- configparser=3.5.0b2=py27_1
- decorator=4.0.9=py27_0
- entrypoints=0.2=py27_1
- fontconfig=2.11.1=5
- freetype=2.5.5=0
- get_terminal_size=1.0.0=py27_0 
- ipykernel=4.3.1=py27_0
- ipython=4.2.0=py27_0
- ipython_genutils=0.1.0=py27_0
- ipywidgets=4.1.1=py27_0
- jinja2=2.8=py27_0
- jsonschema=2.4.0=py27_0
- jupyter=1.0.0=py27_2
- jupyter_client=4.2.2=py27_0
- jupyter_console=4.1.1=py27_0
- jupyter_core=4.1.0=py27_0
- libpng=1.6.17=0
- libsodium=1.0.3=0
- libxml2=2.9.2=0
- markupsafe=0.23=py27_0
- mistune=0.7.2=py27_0
- nbconvert=4.2.0=py27_0
- nbformat=4.0.1=py27_0
- notebook=4.2.0=py27_0
- openssl=1.0.2g=0
- path.py=8.2.1=py27_0
- pexpect=4.0.1=py27_0
- pickleshare=0.5=py27_0
- pip=8.1.1=py27_1
- ptyprocess=0.5=py27_0
- pygments=2.1.3=py27_0
- pyqt=4.11.4=py27_1
- python=2.7.11=0
- pyzmq=15.2.0=py27_0
- qt=4.8.7=1
- qtconsole=4.2.1=py27_0
- readline=6.2=2
- setuptools=20.7.0=py27_0
- simplegeneric=0.8.1=py27_0
- singledispatch=3.4.0.3=py27_0
- sip=4.16.9=py27_0
- six=1.10.0=py27_0
- sqlite=3.9.2=0
- ssl_match_hostname=3.4.0.2=py27_1
- terminado=0.5=py27_1
- tk=8.5.18=0
- tornado=4.3=py27_0
- traitlets=4.2.1=py27_0
- wheel=0.29.0=py27_0
- zeromq=4.1.3=0
- zlib=1.2.8=0
- pip:
  - backports-abc==0.4
  - backports.shutil-get-terminal-size==1.0.0
  - backports.ssl-match-hostname==3.4.0.2
  - ipython-genutils==0.1.0
  - jupyter-client==4.2.2
  - jupyter-console==4.1.1
  - jupyter-core==4.1.0
```

**Comment by takluyver** [comment]
What system are you on? That directory should be created when you log in - it's getting the path from `$XDG_RUNTIME_DIR`.

**Comment by satra** [comment]
on centos 7.2, using a slurm terminal (using `srun --pty bash`)

**Comment by takluyver** [comment]
I would guess that either:

- Whatever is logging you in is setting that environment variable without creating the directory. This is just broken.
- You are doing something that switches from one user to another without clearing environment variables, so it's inheriting the XDG_RUNTIME_DIR from the user you were before. If this is the case, probably the most sensible thing is to unset XDG_RUNTIME_DIR before running the notebook. It will use a fallback.

**Comment by satra** [comment]
@takluyver - thanks for the feedback. 

i know we have upgraded SLURM since i last ran jupyter-notebook on the system, but this was nominally fine before that. it does turn out that the particular directory does not exist when i use `srun` from SLURM to get an interactive terminal but does exist if i simply ssh in.

unsetting the variable works fine. perhaps jupyter-notebook could trap the `OSError` and use the fallback method?

i will check with the SLURM folks as to how they see the variable implemented through an interactive terminal job.

**Comment by satra** [comment]
@takluyver - quick question - has the XDG_RUNTIME_DIR been in use for a while, or was this a recent change (as in within the last 2 months)?

the reason i ask is that i was able to fire up jupyter-notebooks fine within our SLURM environment. and the SLURM mailing list suggested that SLURM does not support XDG.

**Comment by takluyver** [comment]
We have been using XDG_RUNTIME_DIR for some time, since the switch to Jupyter.

It's fine if it's not supported, but if the directory is not created, the environment variable for it should not be set. Setting it implies that whatever was handling the login supported it.

**Comment by satra** [comment]
yes, SLURM carries over local environment variables and since i logged in via ssh to the head node that variable was set. since you have been using it for a while, i have a suspicion that SLURM changed how remote logins were processed. i'm closing this for now. but will post any updates i get from the SLURM folks.

**Comment by takluyver** [comment]
It seems like a misfeature to carry over arbitrary environment variables, as they could well be wrong. However, if that's what it's doing, I think that `unset XDG_RUNTIME_DIR` before starting the notebook should make it work.

**Comment by eburgueno** [comment]
Hello, we just came across this same issue but with a different job scheduler (OpenLava). We reported it in their [mailing list](https://groups.google.com/forum/#!topic/openlava-users/CFaD5pGMfw0) as well.

> It seems like a misfeature to carry over arbitrary environment variables, as they could well be wrong.

I did want to raise it here too however, as passing the entire set of environment variables has been pretty much standard behaviour for most job schedulers out there, AFAIK. In general, it's the SysAdmin job to ensure that all servers in a cluster are consistent so that the environment is portable. There's probably good reasons why job schedulers do not open a login session for each job they run in execution nodes, but that's just a guess.

I am unsure whether this means the XDG specification is broken when working with HPC-type environments, or if there is actually scope for the application (Jupyter Notebook in this case) to be diligent in checking `XDG_RUNTIME_DIR` is indeed writable.

I think falling back to non-XDG behaviour instead of crashing would provide a better user experience.
